### PR TITLE
Fix archlinux

### PR DESCRIPTION
**Pull request by lucasdemarchi**
With these commits I can generate an archlinux image to test kmod again.

### CODE REVIEW

File: mkosi [hunk 1]
```diff
@@ -1496,6 +1492,10 @@ SigLevel    = Required DatabaseOptional
     with open(os.path.join(workspace, 'root', 'etc/locale.conf'), 'w') as f:
         f.write('LANG=en_US.UTF-8\n')
 
+    # At this point, no process should be left running, kill then
+    run(["fuser", "-c", root, "--kill"])
```
Comment (fsateler) [inline mkosi]: on line 1480 you are killing gpg explicitly, and your comment says dirmngr stays alive.Why is that not a bug in gpg?

**Comment by lucasdemarchi** [comment]
> I'm not sure killing everything is the right solution. Anything we start we should be able to stop before completing, no?

We are stopping everything, but pacstrap may start other things even indirectly and they are not being properly stopped.  This is a last measure if anything happened to be left running, because there's no way the unmount that follows this phase will succeed  (and currently it actually silently fails).

File: mkosi (hunk 1 above)
Comment (lucasdemarchi) [reply mkosi]: Are they the same thing? I do see a "bye" as a result from this command. I thought it was another process that was left running.

Comment (keszybz) [reply mkosi]: Seems to be a bug in gpg. Killing stuff like that is too ugly.

Comment (lucasdemarchi) [reply mkosi]: @falconindy any pointer on the pacstrap/pacman side where the fix for this problem would be? For context, it's agout dirmngr left running after those command return.

**Comment by poettering** [comment]
I am not sure I like the first patch, it feels like something pacstrap should really fix.

The second one makes a ton of sense though. Any chance you can split that one off at least so that we can merge it quickly?

**Comment by lucasdemarchi** [comment]
But then I think we should also start checking by the return of umount.

Biggest problems for me are that 1) we already tried and it broke again. I guess it depends on packages you install; 2) we just try to continue assuming umount succeeded and then fail in a unrelated point. 

I considered killing everything as a sanity thing so this doesn't keep  breaking. We can't umount the for anyway.

**Comment by poettering** [comment]
I'd feel more comfortable if we'd instead of killing evreything that access a mount would just kill everything we forked off. specifically, that we modify run() to allocate a new process group for forked off children and then kill that after the main child exits.

**Comment by lucasdemarchi** [comment]
@poettering for all calls to `subprocess.run()` or only the `run_*`() inside `install_arch()`?

**Comment by poettering** [comment]
I think doing the same for every command we call would be pretty nice.

**Comment by poettering** [comment]
(i mean, rpm and dpkg might result in children anyway, if scriptlets do weird stuff and fork something off in the bg, hence i think it's actually a really good idea if we kill it all.)

**Comment by lucasdemarchi** [comment]
so... I think we need to implement our own `run()` to use `subprocess.Popen` rather than `subprocess.run()`. `Popen` has a creationflags argument that accepts CREATE_NEW_PROCESS_GROUP for calling setpgrp().  For me this is the cleaner way to do it. Do you see another one?

**Comment by poettering** [comment]
i am not a python guru. maybe @keszybz has an idea?

**Comment by lucasdemarchi** [comment]
oh, well.. That was added in python 3.7 only :-/

**Comment by keszybz** [comment]
> Changed in version 3.7: Added the `text` parameter, as a more understandable alias of `universal_newlines`. 

Yay, finally.

**Comment by keszybz** [comment]
```pycon
>>> subprocess.Popen('/bin/ls', startupinfo=...)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.7/subprocess.py", line 670, in __init__
    raise ValueError("startupinfo is only supported on Windows "
ValueError: startupinfo is only supported on Windows platforms
```
The docs say so too.

**Comment by lucasdemarchi** [comment]
I implemented using the preexec_fn callable, setting the pgrpid and then later killing all process.

However it does't work. The 2 binaries left running daemonize so killing on the same pgrpid doesn't work.

**Comment by keszybz** [comment]
So my vote would be to accept your original patch then. It's a bit ugly, but meh.

**Comment by keszybz** [comment]
I'll merge this now since it makes arch functional again. We can always merge a more comprehensive solution in the future.

**Pull request merged**